src/arkode/arkode_lsrkstep.c

```diff
@@ -0,0 +1,1515 @@
+
+  if (!(ark_mem->fn_is_current) && ark_mem->initsetup)
+  {
+    retval = step_mem->fe(ark_mem->tn, ark_mem->yn, ark_mem->fn,
+                          ark_mem->user_data);
+    step_mem->nfe++;
+    ark_mem->fn_is_current = SUNTRUE;
+    if (retval != ARK_SUCCESS) { return (ARK_RHSFUNC_FAIL); }
+  }
+
+  /* A tentative solution at t+h is returned in
+     y and its slope is evaluated in temp1.  */
+
+  w0 = (ONE + TWO / (c13 * SUNSQR((sunrealtype)(step_mem->reqstages))));
+
+  temp1 = SUNSQR(w0) - ONE;
+  temp2 = SUNRsqrt(temp1);
+  arg   = step_mem->reqstages * SUNRlog(w0 + temp2);
+
+  w1 = SUNRsinh(arg) * temp1 /
+       (SUNRcosh(arg) * step_mem->reqstages * temp2 - w0 * SUNRsinh(arg));
+
+  bjm1 = ONE / SUNSQR(TWO * w0);
+  bjm2 = bjm1;
+
+  /* Evaluate the first stage */
+  N_VScale(ONE, ark_mem->yn, ark_mem->tempv1);
+
+  mus = w1 * bjm1;
+
+  N_VLinearSum(ONE, ark_mem->yn, ark_mem->h * mus, ark_mem->fn, ark_mem->tempv2);
+
+  thjm2  = ZERO;
+  thjm1  = mus;
+  zjm1   = w0;
+  zjm2   = ONE;
+  dzjm1  = ONE;
+  dzjm2  = ZERO;
+  d2zjm1 = ZERO;
+  d2zjm2 = ZERO;
+
+  /* Evaluate stages j = 2,...,step_mem->reqstages */
+  for (int j = 2; j <= step_mem->reqstages; j++)
+  {
+    zj   = TWO * w0 * zjm1 - zjm2;
+    dzj  = TWO * w0 * dzjm1 - dzjm2 + TWO * zjm1;
+    d2zj = TWO * w0 * d2zjm1 - d2zjm2 + FOUR * dzjm1;
+    bj   = d2zj / SUNSQR(dzj);
+    ajm1 = ONE - zjm1 * bjm1;
+    mu   = TWO * w0 * bj / bjm1;
+    nu   = -bj / bjm2;
+    mus  = mu * w1 / w0;
+
+    /* Use the ycur array for temporary storage here */
+    retval = step_mem->fe(ark_mem->tcur + ark_mem->h * thjm1, ark_mem->tempv2,
+                          ark_mem->ycur, ark_mem->user_data);
+    if (retval != ARK_SUCCESS) { return (ARK_RHSFUNC_FAIL); }
+
+    cvals[0] = mus * ark_mem->h;
+    Xvecs[0] = ark_mem->ycur;
+    cvals[1] = nu;
+    Xvecs[1] = ark_mem->tempv1;
+    cvals[2] = ONE - mu - nu;
+    Xvecs[2] = ark_mem->yn;
+    cvals[3] = mu;
+    Xvecs[3] = ark_mem->tempv2;
+    cvals[4] = -mus * ajm1 * ark_mem->h;
+    Xvecs[4] = ark_mem->fn;
+
+    retval = N_VLinearCombination(step_mem->nfusedopvecs, cvals, Xvecs,
+                                  ark_mem->ycur);
+    if (retval != 0) { return (ARK_VECTOROP_ERR); }
+
+    thj = mu * thjm1 + nu * thjm2 + mus * (ONE - ajm1);
+
+    /* Shift the data for the next stage */
+    if (j < step_mem->reqstages)
+    {
+      /* To avoid two data copies we swap ARKODE's tempv1 and tempv2 pointers*/
+      N_Vector* ptrtempv1 = &(ark_mem->tempv1);
+      N_Vector* ptrtempv2 = &(ark_mem->tempv2);
+
+      N_Vector temp = *ptrtempv2;
+      *ptrtempv2    = *ptrtempv1;
+      *ptrtempv1    = temp;
+
+      N_VScale(ONE, ark_mem->ycur, ark_mem->tempv2);
+
+      thjm2  = thjm1;
+      thjm1  = thj;
+      bjm2   = bjm1;
+      bjm1   = bj;
+      zjm2   = zjm1;
+      zjm1   = zj;
+      dzjm2  = dzjm1;
+      dzjm1  = dzj;
+      d2zjm2 = d2zjm1;
+      d2zjm1 = d2zj;
+    }
+  }
+  step_mem->nfe += step_mem->reqstages - 1;
```

Comment:
This counter should be incremented in the loop so that if an error occurs before this point it's accurate.